stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -659,10 +659,20 @@ end
 # disambiguation methods: * of Diagonal and Adj/Trans AbsVec
 *(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal) = Adjoint(map((t,s) -> t'*s, D.diag, parent(x)))
 *(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal) = Transpose(map((t,s) -> transpose(t)*s, D.diag, parent(x)))
-*(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
-    mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
-*(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
-    mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
+function *(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector)
+    if all(isempty(x), isempty(D), isempty(y))
+        return zero(promote_type(eltype(x), eltype(D), eltype(y)))
+    else
+        return mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
+    end
+end
+function *(x::Transpose{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector)
+    if all(isempty(x), isempty(D), isempty(y))
+        return zero(promote_type(eltype(x), eltype(D), eltype(y)))
+    else
+        return mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))
+    end
+end
 function dot(x::AbstractVector, D::Diagonal, y::AbstractVector)
     mapreduce(t -> dot(t[1], t[2], t[3]), +, zip(x, D.diag, y))
```

Comment:
I guess this function can use the same implementation too? Also, since now we have a non-trivial function body, maybe it makes sense to define a separate internal function and call it from the above three functions?

Reply:
Sounds good. I called the the internal function `_dot` but that's not a great name. Maybe it should be called something like `_mul` or even `_mapreduce_mul`?

Reply:
I think `_dot` is good.  It's a unique name in LinearAlgebra so it's easy to rename afterwards if we need to.